### PR TITLE
Feature/add matchsrnn model

### PR DESCRIPTION
**Pull request by dyuyang**

### CODE REVIEW

File: matchzoo/modules/matching_tensor.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                 init_diag: bool = True):
+        """:class:`MatchingTensor` constructor."""
+        super().__init__()
+        self._matching_dims = matching_dims
+        self._channels = channels
+        self._normalize = normalize
+        self._init_diag = init_diag
+
+        # Used purely for shape validation.
+        if not isinstance(matching_dims, list) or len(matching_dims) != 2:
```
Comment (Chriskuei) [inline matchzoo/modules/matching_tensor.py]: matching_dims can be `int`

File: matchzoo/modules/matching_tensor.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        if self._init_diag:
+            interaction_matrix = np.float32(
+                np.random.uniform(
+                    -0.05, 0.05,
+                    [self._channels, self._matching_dims[0], self._matching_dims[1]]
+                )
+            )
+            for channel_index in range(self._channels):
+                np.fill_diagonal(interaction_matrix[channel_index], 0.1)
```
Comment (Chriskuei) [inline matchzoo/modules/matching_tensor.py]: Use `torch.fill_diagonal`

File: matchzoo/modules/matching_tensor.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            self.interaction_matrix = nn.Parameter(torch.tensor(interaction_matrix))
+        else:
+            self.interaction_matrix = nn.Parameter(torch.empty(
+                [self._channels, self._matching_dims[0], self._matching_dims[1]]
+            ).uniform_())
+
+    def forward(self, inputs):
```
Comment (Chriskuei) [inline matchzoo/modules/matching_tensor.py]:
```suggestion
    def forward(self, x, y):
```

File: matchzoo/modules/spatial_gru.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        return h_ij
+
+    def forward(self, inputs):
+        """
+        Perform SpatialGRU on word interation matrix.
+
+        :param inputs: input tensors.
+        """
+
+        batch_size, channels, left_maxlen, right_maxlen = inputs.shape
```
Comment (Chriskuei) [inline matchzoo/modules/spatial_gru.py]:
```suggestion
        batch_size, channels, left_length, right_length = inputs.shape
```

File: matchzoo/models/match_srnn.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        ))
+        return params
+
+    def build(self):
+        """Build model structure."""
+
+        self.embedding = self._make_default_embedding_layer()
+        self.dropout = nn.Dropout(p=self._params['dropout'])
+
+        self.matching_tensor_network = MatchingTensor(
```
Comment (caiyinqiong) [inline matchzoo/models/match_srnn.py]: rename `matching_tensor_network` to `matching_tensor`.

File: matchzoo/models/match_srnn.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            self._params['embedding_output_dim'],
+            channels=self._params["channels"])
+
+        self.spatial_gru_network = SpatialGRU(
```
Comment (caiyinqiong) [inline matchzoo/models/match_srnn.py]: rename `spatial_gru_network` to `spatial_gru`.

File: matchzoo/modules/spatial_gru.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        >>> channels, units= 4, 10
+        >>> spatial_gru = mz.modules.SpatialGRU(channels, units)
+
+    """
+
+    def __init__(
+        self,
+        channels: int = 4,
+        units: int = 10,
+        activation: nn.Module = nn.Tanh,
```
Comment (caiyinqiong) [inline matchzoo/modules/spatial_gru.py]: activation should support `str` or `nn.Module` instance.

**Pull request merged**